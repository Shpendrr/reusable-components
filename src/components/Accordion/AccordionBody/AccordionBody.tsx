import React, { FC, memo } from "react";

export interface IAccordionBodyProps extends React.HTMLAttributes<HTMLDivElement> {
  open?: boolean;
  children: React.ReactNode;
}

const AccordionBody: FC<IAccordionBodyProps> = ({ open, children, ...rest }) => {
  const bodyStyle = `
    ${open ? "h-auto" : "h-0"}
    ${open ? "visible overflow-visible scale-y-100" : "invisible overflow-hidden scale-y-0"}
    transition-transform transition-height duration-200
  `;

  return (
    <div className={`p-4 w-full lg:w-[65%] lg:mx-auto ${bodyStyle}`} {...rest}>
      {children}
    </div>
  );
};

export default memo(AccordionBody);