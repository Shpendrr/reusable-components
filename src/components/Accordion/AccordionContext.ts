import { createContext, useContext } from "react"


export interface IAccordionContext {
    localOpen: boolean;
    setLocalOpen?: (value: boolean) => void;
    handleToggle: () => void;
}


export const defaultValues: IAccordionContext = {
    localOpen: false,
    setLocalOpen: () => { },
    handleToggle: () => { }
}



export const AccordionContext = createContext<IAccordionContext>(defaultValues);


export const useAccordionContext = (): IAccordionContext => useContext(AccordionContext);

