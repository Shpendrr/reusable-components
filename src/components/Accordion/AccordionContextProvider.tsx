import React, { FC, useCallback, useState } from "react";
import { AccordionContext, IAccordionContext } from "./AccordionContext";

export interface IAccordionContextProvider {
  children: React.ReactNode;
}

const AccordionContextProvider: FC<IAccordionContextProvider> = ({
  children,
}) => {
  const [localOpen, setLocalOpen] = useState<boolean>(false);

  const handleToggle = useCallback(() => {
    setLocalOpen((prev) => !prev);
  }, [localOpen]);

  const ContextValues: IAccordionContext = {
    localOpen,
    setLocalOpen,
    handleToggle,
  };
  return (
    <AccordionContext.Provider value={ContextValues}>
      {children}
    </AccordionContext.Provider>
  );
};

export default AccordionContextProvider;
