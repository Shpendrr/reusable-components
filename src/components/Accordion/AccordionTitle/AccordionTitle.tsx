import Button from "@/components/Button/Button";
import React, { ButtonHTMLAttributes, memo } from "react";
import cn from "classnames";
import { useAccordionContext } from "../AccordionContext";

export interface AccordionTitleProps
  extends ButtonHTMLAttributes<HTMLButtonElement> {
  open?: boolean;
}

const AccordionTitle: React.FC<AccordionTitleProps> = ({
  children,
  open,
  ...rest
}) => {
  const { handleToggle } = useAccordionContext();

  return (
    <Button
      className="w-full lg:w-[65%] mt-[50px] min-h-48 flex justify-between items-center p-4 border-[#d7d8df] border-b-2 bg-transparent text-base lg:mx-auto"
      onClick={handleToggle}
      {...rest}
    >
      {children}
    </Button>
  );
};

export default memo(AccordionTitle);