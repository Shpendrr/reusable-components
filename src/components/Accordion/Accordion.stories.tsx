import React from "react";
import Accordion, { AccordionProps } from "@/components/Accordion/Accordion";
import AccordionTitle from "@/components/Accordion/AccordionTitle/AccordionTitle";
import AccordionBody from "@/components/Accordion/AccordionBody/AccordionBody";
import { Meta } from "@storybook/react";

export default {
  title: "Components/Accordion",
  component: Accordion,
} as Meta;

const Template = (args: React.JSX.IntrinsicAttributes & AccordionProps) => (
  <Accordion {...args}>
    <AccordionTitle>Accordion Title</AccordionTitle>
    <AccordionBody>Accordion Body</AccordionBody>
  </Accordion>
);

export const Default = Template.bind({});
