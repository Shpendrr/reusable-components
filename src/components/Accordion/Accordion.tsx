import React, { memo, useState, useMemo, Children, useCallback } from "react";
import cn from "classnames";
import AccordionTitle from "@/components/Accordion/AccordionTitle/AccordionTitle";
import AccordionBody from "@/components/Accordion/AccordionBody/AccordionBody";
import { useAccordionContext } from "./AccordionContext";

export interface AccordionProps extends React.HTMLAttributes<HTMLDivElement> {
  children: React.ReactNode;
  open?: boolean;
}

function Accordion(props: AccordionProps): React.ReactElement {
  const { children, open, ...rest } = props;

  const { handleToggle, localOpen } = useAccordionContext();

  const processedChildren = useMemo(() => {
    return Children.map(children, (child, index) => {
      if (!React.isValidElement(child)) return child;

      const isOpen = open !== undefined ? open : localOpen;

      if (index === 0) {
        return <AccordionTitle {...child.props} open={isOpen} />;
      }

      if (index === 1) {
        return <AccordionBody {...child.props} open={isOpen} />;
      }

      return child;
    });
  }, [children, localOpen, open, handleToggle]);

  return (
    <div className={cn("w-full")} {...rest}>
      {processedChildren}
    </div>
  );
}

export default memo(Accordion);
