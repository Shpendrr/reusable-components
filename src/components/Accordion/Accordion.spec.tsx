import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Accordion from "./Accordion";
import AccordionContextProvider from "@/components/Accordion/AccordionContextProvider";
import AccordionTitle from "@/components/Accordion/AccordionTitle/AccordionTitle";
import AccordionBody from "@/components/Accordion/AccordionBody/AccordionBody";
import "@testing-library/jest-dom";

describe("Accordion", () => {
  it("should render the title and body components", () => {
    const { getByText } = render(
      <AccordionContextProvider>
        <Accordion>
          <AccordionTitle>Accordion Title</AccordionTitle>
          <AccordionBody>Accordion Body</AccordionBody>
        </Accordion>
      </AccordionContextProvider>
    );

    expect(getByText("Accordion Title")).toBeInTheDocument();
    expect(getByText("Accordion Body")).toBeInTheDocument();
  });

  it("should toggle the accordion body when the title is clicked", () => {
    const { getByText } = render(
      <AccordionContextProvider>
        <Accordion>
          <AccordionTitle>Accordion Title</AccordionTitle>
          <AccordionBody>Accordion Body</AccordionBody>
        </Accordion>
      </AccordionContextProvider>
    );

    const title = getByText("Accordion Title");
    const body = getByText("Accordion Body");

    expect(body).not.toBeVisible();

    fireEvent.click(title);

    expect(body).toBeVisible();

    fireEvent.click(title);

    expect(body).not.toBeVisible();
  });
  it("should match the snapshot", () => {
    const { container } = render(
      <AccordionContextProvider>
        <Accordion>
          <AccordionTitle>Accordion Title</AccordionTitle>
          <AccordionBody>Accordion Body</AccordionBody>
        </Accordion>
      </AccordionContextProvider>
    );

    expect(container).toMatchSnapshot();
  });
});
