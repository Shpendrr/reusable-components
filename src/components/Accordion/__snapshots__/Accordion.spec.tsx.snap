// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should match the snapshot 1`] = `
<div>
  <div
    class="w-full"
  >
    <button
      class="w-full lg:w-[65%] mt-[50px] min-h-48 flex justify-between items-center p-4 border-[#d7d8df] border-b-2 bg-transparent text-base lg:mx-auto"
    >
      Accordion Title
    </button>
    <div
      class="p-4 w-full lg:w-[65%] lg:mx-auto h-0 invisible overflow-hidden scale-y-0 transition-transform transition-height duration-200"
    >
      Accordion Body
    </div>
  </div>
</div>
`;
